### ISSUE TITLE
HTTP Headers capture feature resulting into 100% CPU & Stuck Threads

### ISSUE THREAD

**Issue by zoomingrocket** [microsoft/ApplicationInsights-Java#2219]
### Expected behavior
HTTP Headers to be captured as defined in json configuration without any 100% CPU usage or stuck threads

### Actual behavior
High CPU usage and stuck threads

### To Reproduce
COTs Middleware system, upon system restart we start seeing repeat error stacks in Application Log as below
```
Apr 7, 2022 3:34:34 PM UTC INFO    [org.eclipse.jetty.client.ResponseNotifier notifySuccess] Exception while notifying listener org.eclipse.jetty.client.HttpRequest$13@2de8889e
java.lang.reflect.UndeclaredThrowableException
        at com.sun.proxy.$Proxy100.onSuccess(Unknown Source)
        at org.eclipse.jetty.client.ResponseNotifier.notifySuccess(ResponseNotifier.java:147)
        at org.eclipse.jetty.client.ResponseNotifier.notifySuccess(ResponseNotifier.java:139)
        at org.eclipse.jetty.client.HttpReceiver.responseSuccess(HttpReceiver.java:391)
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.messageComplete(HttpReceiverOverHTTP.java:266)
        at org.eclipse.jetty.http.HttpParser.parseContent(HttpParser.java:1404)
        at org.eclipse.jetty.http.HttpParser.parseNext(HttpParser.java:1246)
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.parse(HttpReceiverOverHTTP.java:156)
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.process(HttpReceiverOverHTTP.java:117)
        at org.eclipse.jetty.client.http.HttpReceiverOverHTTP.receive(HttpReceiverOverHTTP.java:69)
        at org.eclipse.jetty.client.http.HttpChannelOverHTTP.receive(HttpChannelOverHTTP.java:89)
        at org.eclipse.jetty.client.http.HttpConnectionOverHTTP.onFillable(HttpConnectionOverHTTP.java:123)
        at org.eclipse.jetty.io.AbstractConnection$2.run(AbstractConnection.java:544)
        at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:635)
        at org.eclipse.jetty.util.thread.QueuedThreadPool$3.run(QueuedThreadPool.java:555)
        at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: java.lang.reflect.InvocationTargetException
        at jdk.internal.reflect.GeneratedMethodAccessor98.invoke(Unknown Source)
        at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
        at java.base/java.lang.reflect.Method.invoke(Method.java:566)
        at io.opentelemetry.javaagent.shaded.instrumentation.jetty.httpclient.v9_2.internal.JettyClientWrapUtil.lambda$wrapTheListener$1(JettyClientWrapUtil.java:69)
        ... 16 more
Caused by: java.lang.NoClassDefFoundError: Could not initialize class io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpHeaderAttributes
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpCommonAttributesExtractor.onEnd(HttpCommonAttributesExtractor.java:79)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpClientAttributesExtractor.onEnd(HttpClientAttributesExtractor.java:54)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.Instrumenter.end(Instrumenter.java:224)
        at io.opentelemetry.javaagent.shaded.instrumentation.jetty.httpclient.v9_2.internal.JettyHttpClient9TracingInterceptor.closeIfPossible(JettyHttpClient9TracingInterceptor.java:172)
        at io.opentelemetry.javaagent.shaded.instrumentation.jetty.httpclient.v9_2.internal.JettyHttpClient9TracingInterceptor.onSuccess(JettyHttpClient9TracingInterceptor.java:152)
        at org.eclipse.jetty.client.HttpRequest$13.onSuccess(HttpRequest.java:537)
        ... 20 more
```

#### Sample Application
NA, COTs Middleware system serving HTTP APIs using Jetty, etc. tech stack

### System information
Please provide the following information:
 - SDK Version: Java Agent 3.2.10
 - OS type and version: CentOS 7.9
 - Application Server type and version (if applicable): 
 - Using spring-boot? No
 - Additional relevant libraries (with version, if applicable):
 Java Version :
openjdk version "11.0.14" 2022-01-18 LTS
OpenJDK Runtime Environment Corretto-11.0.14.9.1 (build 11.0.14+9-LTS)

### Logs
No Errors in Application Insight Agent Logs
Thread Dump of Application results in 100s of below threads
```
"LocalProcessExec-thread-33" Id=3577 RUNNABLE
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.Instrumenter.end(Instrumenter.java:224)
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
        -  locked sun.net.www.protocol.https.DelegateHttpsURLConnection@79d26ef2
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:3099)
        at io.opentelemetry.javaagent.instrumentation.httpurlconnection.HttpUrlHttpAttributesExtractor.responseHeader(HttpUrlHttpAttributesExtractor.java:75)
        at io.opentelemetry.javaagent.instrumentation.httpurlconnection.HttpUrlHttpAttributesExtractor.responseHeader(HttpUrlHttpAttributesExtractor.java:17)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpCommonAttributesExtractor.onEnd(HttpCommonAttributesExtractor.java:77)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpClientAttributesExtractor.onEnd(HttpClientAttributesExtractor.java:54)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.Instrumenter.end(Instrumenter.java:224)
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
        -  locked sun.net.www.protocol.https.DelegateHttpsURLConnection@79d26ef2
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:3099)
        at io.opentelemetry.javaagent.instrumentation.httpurlconnection.HttpUrlHttpAttributesExtractor.responseHeader(HttpUrlHttpAttributesExtractor.java:75)
        at io.opentelemetry.javaagent.instrumentation.httpurlconnection.HttpUrlHttpAttributesExtractor.responseHeader(HttpUrlHttpAttributesExtractor.java:17)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpCommonAttributesExtractor.onEnd(HttpCommonAttributesExtractor.java:77)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpClientAttributesExtractor.onEnd(HttpClientAttributesExtractor.java:54)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.Instrumenter.end(Instrumenter.java:224)
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
        -  locked sun.net.www.protocol.https.DelegateHttpsURLConnection@79d26ef2
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getHeaderField(HttpURLConnection.java:3099)
        at io.opentelemetry.javaagent.instrumentation.httpurlconnection.HttpUrlHttpAttributesExtractor.responseHeader(HttpUrlHttpAttributesExtractor.java:75)
        at io.opentelemetry.javaagent.instrumentation.httpurlconnection.HttpUrlHttpAttributesExtractor.responseHeader(HttpUrlHttpAttributesExtractor.java:17)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpCommonAttributesExtractor.onEnd(HttpCommonAttributesExtractor.java:77)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.http.HttpClientAttributesExtractor.onEnd(HttpClientAttributesExtractor.java:54)
        at io.opentelemetry.javaagent.shaded.instrumentation.api.instrumenter.Instrumenter.end(Instrumenter.java:224)
        at java.base@11.0.14/sun.net.www.protocol.http.HttpURLConnection.getInputStream(HttpURLConnection.java:1520)
```

### Screenshots
![image](https://user-images.githubusercontent.com/2420449/162249831-ce432d0a-a3da-4e5d-a800-bf509fb9784e.png)

**Comment by trask** [comment]
Hi @zoomingrocket!

can you try with the latest SNAPSHOT build?

https://github.com/microsoft/ApplicationInsights-Java/suites/5977750973/artifacts/205795109

The `HttpUrlConnection` instrumentation looks like it's going into an infinite loop, and that instrumentation has been reworked recently, so I'm hoping that may resolve your issue.

If it doesn't resolve the issue, can you attach full stack trace from when using the SNAPSHOT build? (I'm assuming you're getting a `StackOverflowError`?)

**Comment by zoomingrocket** [comment]
@trask  - I am trying out the latest snapshot over the weekend and will confirm by Monday EOD

**Comment by trask** [comment]
@zoomingrocket any luck? looking forward to any feedback as looks like this is affecting someone else also in #2218

**Comment by zoomingrocket** [comment]
@trask - Snapshot build does fix the issue. Over the last 6 days, CPU utilization has been minimal and pretty stable including server load averages.
HTTP Headers are captured accurately in custom props as requested in json configuration

![image](https://user-images.githubusercontent.com/2420449/164358821-40d7d2e6-a551-4abd-99e5-8ed0c91ce44d.png)

**Comment by zoomingrocket** [comment]
@trask  - We do have non-related new error in the application insight log, not related to the original issue, maybe due to snapshot build? Let me know if we should open a separate issue for this

```
2022-04-21 02:23:18.775Z WARN  c.a.m.o.e.i.p.TelemetryPipeline - In the last 5 minutes, the following operation has failed 3 times (out of 262): Sending telemetry to the ingestion service:
 * 109: Field 'message' on type 'MessageData' is required but missing or empty. Expected: string, Actual:
 (3 times)
```

**Comment by trask** [comment]
@zoomingrocket that's great to hear!

and thanks for reporting this additional issue! I can reproduce that as well and will get fix in soon.

**Comment by trask** [comment]
this additional issue is fixed now (in 3.3.0-SNAPSHOT) by #2245
